[Inductor] Finely overlap template fusion with async_compile

PaulZhang12 opened the pull request:
Differential Revision: D89255518

Currently during epilogue fusion evaluation, the epilogues that get benchmarked/statically analyzed are based on the order of nodes coming into the fusion. However, this does not take full advantage of the async compilation of templates with epilogues/prologues. 

We fix that by extracting the compile futures from the call to speedup_fn. We then take the futures that first complete, still maintaining the template fusion order, where epilogues/prologues that can be fused into multiple templates. We also cache epilogue/prologue fusions that have been evaluated previously, as repeated codegen IO for the template fusions are expensive.

Two main optimizations are introduced for max-autotune compile time in this PR:
1. Cache template fusions that have already been seen. While template fusions that have been previously seen already aren't benchmarked, the cost of writing each fusion to disk introduces overhead, especially with repeated ones over many rounds of fusion. 
2. Kick off all epilogue/prologue kernel compilation in parallel, which takes advantage of async_compile process pool for efficient parallelization. Evaluate epilogue and prologue fusions based on original fusion order for each epilogue/prologue candidate as compilation is done through evaluating the futures.

## **Compile time experiments all done with representative internal arch**
Baseline -> ~12s

**Standard Max-Autotune**
Max-autotune - 144s
Max-autotune - overlap template fusions -> 100s

**All with diode topk=1**
Max-autotune + no epilogue/prologue fusions -> ~18s
Max-autotune + benchmarking epilogues -> ~59s
Max-autotune + benchmarking epilogues + overlap template fusions -> ~29s
* without caching: 46s
* without overlapping: 54s

Max-autotune + static analyzing epilogues -> ~30s
Max-autotune + static analyzing epilogues + async pipelined autotuning -> 47s
Max-autotune + static analyzing epilogues + overlap template fusions + async pipelined autotuning -> ~18s

## **TorchInductor CI compile time**
Below is data collected from manually running torchbench locally on an H100 machine, as the CI compilation numbers can often be flaky/unreliable.


TIMM | compilation_latency | compilation_latency optimization | Compilation Latency Increase (s)
-- | -- | -- | --
adv_inception_v3 | 112.349167 | 112.760696 | 0.411529
beit_base_patch16_224 | 95.903498 | 97.692569 | 1.789071
convnextv2_nano.fcmae_ft_in22k_in1k | 118.273649 | 121.011309 | 2.73766
deit_base_distilled_patch16_224 | 80.767883 | 81.779595 | 1.011712
deit_tiny_patch16_224.fb_in1k | 70.213287 | 70.691566 | 0.478279
dm_nfnet_f0 | 115.08765 | 111.540226 | -3.547424
ghostnet_100 | 171.751849 | 169.028063 | -2.723786
inception_v3 | 168.096137 | 171.589615 | 3.493478
mobilenetv2_100 | 131.443499 | 134.202953 | 2.759454
mobilenetv3_large_100 | 232.691377 | 241.775705 | 9.084328
mobilevit_s | 162.556104 | 157.611769 | -4.944335
nfnet_l0 | 106.444117 | 103.529731 | -2.914386
repvgg_a2 | 89.10137 | 92.640802 | 3.539432
swin_base_patch4_window7_224 | 230.605837 | 232.994867 | 2.38903
tf_efficientnet_b0 | 131.88646 | 133.045255 | 1.158795
visformer_small | 87.243081 | 86.116177 | -1.126904
vit_base_patch16_siglip_256 | 0 | 0 | 0



HuggingFace | compilation_latency | compilation_latency optimization | Compilation Latency Increase (s)
-- | -- | -- | --
AlbertForMaskedLM | 91.178588 | 44.86694 | -46.311648
AllenaiLongformerBase | 128.938647 | 125.305844 | -3.632803
BartForCausalLM | 22.288702 | 21.940061 | -0.348641
BertForMaskedLM | 95.562674 | 89.870288 | -5.692386
BlenderbotForCausalLM | 102.171161 | 101.956513 | -0.214648
DebertaV2ForMaskedLM | 155.576169 | 153.626961 | -1.949208
DistilBertForMaskedLM | 81.940878 | 84.612052 | 2.671174
DistillGPT2 | 73.420599 | 70.765531 | -2.655068
ElectraForCausalLM | 42.653002 | 43.163519 | 0.510517
GPT2ForSequenceClassification | 101.844608 | 108.868427 | 7.023819
GoogleFnet | 75.241814 | 77.868139 | 2.626325
LayoutLMForMaskedLM | 95.432125 | 99.822281 | 4.390156
M2M100ForConditionalGeneration | 46.549122 | 46.859677 | 0.310555
MBartForCausalLM | 24.602922 | 24.559812 | -0.04311
MT5ForConditionalGeneration | 99.856653 | 102.516542 | 2.659889
MegatronBertForCausalLM | 128.070552 | 126.370065 | -1.700487
MobileBertForMaskedLM | 141.329823 | 134.774825 | -6.554998
OPTForCausalLM | 34.37166 | 33.30809 | -1.06357
PLBartForCausalLM | 31.595436 | 30.708169 | -0.887267
PegasusForCausalLM | 22.28014 | 21.993263 | -0.286877
Qwen/Qwen3-0.6B | 0 | 0 | 0
RobertaForCausalLM | 99.922612 | 96.397972 | -3.52464
T5ForConditionalGeneration | 72.263815 | 73.285807 | 1.021992
T5Small | 71.761801 | 73.808848 | 2.047047
TrOCRForCausalLM | 27.9683 | 28.123752 | 0.155452
XGLMForCausalLM | 29.021936 | 29.015994 | -0.005942
XLNetLMHeadModel | 153.020033 | 147.014865 | -6.005168
YituTechConvBert | 121.180943 | 121.468483 | 0.28754
google/gemma-2-2b | 0 | 0 | 0
meta-llama/Llama-3.2-1B | 0 | 0 | 0
openai/whisper-tiny | 0 | 0 | 0



## **Dashboard for Performance**

**Note: Compilation time for max-autotune is very noisy. Refer to above local results for more accurate compilation time overheads**
Inference:
<img width="1728" height="915" alt="Screenshot 2026-01-12 at 11 17 15 AM" src="https://github.com/user-attachments/assets/bbb7a689-298d-4b60-bc12-fb66ea20e266" />

Training:
<img width="1728" height="677" alt="Screenshot 2026-01-12 at 11 18 04 AM" src="https://github.com/user-attachments/assets/fe46a765-b04a-417e-84e7-2a1ae89c0c3f" />



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -4153,24 +4290,73 @@ def resolve_pending_fusions(
 
+            if (
+                is_template_fusion(node1, node2)
+                and (node1, node2) in self.seen_template_fusions
```
We allow more reordering on the final round. See: https://github.com/pytorch/pytorch/blob/d74f9ecce649b52c52bea857c1c7d1e859ff6826/torch/_inductor/scheduler.py#L4167-L4171

Not relevant for template, maybe relevant for epilogue. cc @shunting314 loop reordering

eellison replied:
it means the cache might be outdated

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -4145,6 +4186,102 @@ def resolve_pending_fusions(
+                template_futures = []
+                future_to_pending_fusion = {}
+
+                remaining_nodes = fusion_nodes - finished_fusions
```
This is fusion_nodes x finished_fusions complexity. anyway to do this differently? fusion_nodes is potentially large.

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -4145,6 +4186,102 @@ def resolve_pending_fusions(
+                for fusion_node in remaining_nodes:
+                    # Pointwise node already fused
+                    if fusion_node not in pending_fusions:
+                        finished_fusions.add(fusion_node)
```
just for my understanding - what if instead of iterating like `for futures, is_speedup_fn, node_key1, node_key2 in pending_fusions.values():` in the original code, we iterated as 

`for fut in as_completed(pending_fusion_futures):`

does this accomplish the same thing as this function does or what am i missing

PaulZhang12 replied:
Hmm we potentially could? Actually the reason I did it this way was to make it more safe and preserve the original ordering.

Imagine we have an epilogue node that can be fused into two template nodes, node1 and node2. In the previous code, node1 + epilogue is evaluated, and if it fails, we go into node2 + epilogue. If we only awaited the futures like that, then node2 + epilogue could come first, which might not be intended

PaulZhang12 replied:
Right, I guess what I mean is during the first iteration through all nodes, we don't actually want to evaluate the fusion just yet. We just want to save the future for later instead?

If we just collect all the available futures in the first loop and in this loop, just loop through all the possible futures, it doesn't preserve the initial order? This is my mental model for these two loops:

1. Collect the first future for each epilogue/prologue fusion. This helps us preserve the original order and not do extra work by evaluating futures for nodes already fused
2. Await futures as_completed parallelization

eellison reviewed (commented):
cool! couple comments

PaulZhang12 commented on torch/_inductor/scheduler.py:
```diff
@@ -3757,7 +3773,7 @@ def _any_atomic_add(self, node_list: Sequence[BaseSchedulerNode]) -> bool:
     def speedup_by_fusion(
         self, node1: BaseSchedulerNode, node2: BaseSchedulerNode
-    ) -> Union[bool, Callable[[], bool]]:
+    ) -> Union[bool, tuple[Callable[[], bool], Optional[LambdaFuture]]]:
```
Sure! I can do this, good suggestion

PaulZhang12 commented on torch/_inductor/scheduler.py:
```diff
@@ -4145,6 +4186,102 @@ def resolve_pending_fusions(
 
                 fuse_two_nodes(node_key1, node_key2)
 
+        def evaluate_pending_template_fusions(
```
It was a bit more convenient to inline given the fact that we use pending_fusions and call fuse_to_nodes, which are both in this scope here. I can move this out though and pass in the pending_fusions as an argument?

shunting314 commented on test/inductor/test_max_autotune.py:
```diff
@@ -3378,6 +3378,73 @@ def foo(x, y, z):
+            return normalized
+
+        if use_async_compile:
+            async_compile = torch._inductor.async_compile.AsyncCompile()
```
nit: call this class method without instantiation

shunting314 commented on torch/_inductor/scheduler.py:
```diff
@@ -4158,12 +4223,117 @@ def benchmark_when_ready() -> bool:
 
-            return benchmark_when_ready
+            return FusionResult.from_callable(
+                callable_fn=benchmark_when_ready, future=future_and_mod_l1_fused[0]
```
Similar question: ideally you want to a future which is ready when all three of future_and_mod_l1/future_and_mod_l2/future_and_mod_l1_fused are ready?

PaulZhang12 replied:
Yeah we could, but back to my previous point, this is more of a "generality" that if one of the compilations is done, the others should be done given that they occur in parallel processes. Here, we assume the compilation of the fusion is at least more expensive than the compilation without the fusion

shunting314 commented on torch/_inductor/scheduler.py:
```diff
@@ -4185,30 +4355,23 @@ def fuse_nodes_once(
         pending_fusions: dict[
             BaseSchedulerNode,
-            tuple[Callable[[], bool], BaseSchedulerNode, BaseSchedulerNode],
+            list[PendingFusion],
```
Why change to a list here?

PaulZhang12 replied:
A given epilogue might be fusible with multiple potential templates (see unit test). In this case, we want to evaluate a list as all the potential template fusions for the epilogue in the later stage where we evaluate template fusions. That way, if a template fusion is not profitable, we try all options, and in the original order

PaulZhang12 commented on torch/_inductor/scheduler.py:
```diff
@@ -4078,7 +4141,9 @@ def benchmark_when_ready() -> bool:
 
-            return benchmark_when_ready
+            return FusionResult.from_callable(
+                benchmark_when_ready, future_choices[0][1]
```
The future is just to evaluate when the fused kernel compilation is complete. If there are multiple futures, they still get benchmarked in `benchmark_when_ready`. The assumption here is that when one of the template fusion compilations is done, the others are pretty much done.

PaulZhang12 commented on torch/_inductor/scheduler.py:
```diff
@@ -4266,7 +4474,15 @@ def resolve_pending_fusions(
+        )
+        self.evaluate_pending_template_fusions(
+            prologue_fusion_nodes, seen_pair_speedup_fn, fused_nodes, pending_fusions
+        )
```
Yes that's right. This is based on the assumption that template fusion is done after other fusions, with a generally lower score

shunting314 replied:
We have this config: https://github.com/pytorch/pytorch/blob/6c619f0f051db120565f7c386bc6ac04067dfecd/torch/_inductor/config.py#L263 to control if we do epilogue fusion first. With the PR, if someone enables the flag, they'll be surprised that it does not take effect.

I'm not sure if there is still anyone using it. cc @eellison  . I think we should at least remove this config to avoid confusion if nobody is using it.

eellison replied:
tbh, i'm not really sure when someone would set this... i'm fine not maintaining this config. could just be for testing.

eellison reviewed (commented):
few more comments - maybe we should VC and discuss

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -4158,12 +4223,119 @@ def benchmark_when_ready() -> bool:
+                    pending_fusion.callable_fn,
+                )
+
+            remaining_nodes = remaining_nodes - fusions_to_remove
```
i suppose we have `fusions_to_remove` bc we're iterating over remaining nodes

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -4158,12 +4223,119 @@ def benchmark_when_ready() -> bool:
+                # speedup already evaluated, don't fuse as wasn't fused prior or
+                # the template node has already been fused
+                if (
+                    pending_fusion.callable_fn in seen_pair_speedup_fn
```
shouldn't this be handled by the caching ?

PaulZhang12 replied:
Yeah I think so, I can remove the callable_fn part

eellison reviewed (commented):
looks good, one last round of comments

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -4158,12 +4223,119 @@ def benchmark_when_ready() -> bool:
+
+    def evaluate_pending_template_fusions(
+        self,
+        fusion_nodes: OrderedSet[BaseSchedulerNode],
```
Nit: below, in the main scheduling loop, we have `fused_nodes` which is the set of all nodes. this is just the list of possible template fusion nodes. this was part of my confusion that we were doing O(all_nodes) operations. 

Could we rename this ? maybe `template_fusion_candidates` or something ? up to you

PaulZhang12 commented on torch/_inductor/scheduler.py:
```diff
@@ -4158,12 +4223,119 @@ def benchmark_when_ready() -> bool:
+                # the template node has already been fused
+                if (
+                    pending_fusion.callable_fn in seen_pair_speedup_fn
+                    or template_node not in fused_nodes
```
`fused_nodes` is the set of all nodes, so `template_node`, if it's no longer in `fused_nodes`, was already fused previously with a prologue/epilogue

eellison replied:
can you add a comment that this doesn't handle the case where we have multiple separate epilogues, but that this shouldn't exist today ?

PaulZhang12 commented on torch/_inductor/scheduler.py:
```diff
@@ -4239,25 +4433,39 @@ def resolve_pending_fusions(
-            if is_speedup_fn in seen_pair_speedup_fn:
+
+        # Resolve pending fusions for non templates
+        for pending_fusion in pending_fusions.values():
```
Yeah exactly, for the benchmark fusion case. This evaluates everything that is not a template fusion

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -4160,12 +4211,118 @@ def benchmark_when_ready() -> bool:
+                    future_to_pending_fusion[f] = (pending_fusion, candidate)
+                else:
+                    # Non AsyncCompile path, perform fusion
+                    if self.fuse_if_speedup(
```
Sorry, when would this be the case ?

eellison reviewed (approved):
Looks good. Just to discuss a bit on how we got here for posterity. Assumption: prologue fusions are rare but should still be handled.

 The simpler, straw man proposal is:

- add the caching

Do [this loop](https://github.com/pytorch/pytorch/blob/f2c5f9f66bef660d3b9f846bc53e993863a4b39b/torch/_inductor/scheduler.py#L4261-L4263) as 

```
for is_speedup_fn, node_key1, node_key2 in `as_completed(pending_fusions.values())`:
```

As you said, typically, a template should only have a single epilogue, since the epilogues themselves should have already fused. So I think this simpler version would already account for 90% of the speedups. The remaining case that your pr is trying to tackle is when an epilogue has multiple possible templates it could fuse into. It would be awesome if you could verify that this simpler version doesn’t regress performance on what you’re looking at before landing. 

One last comment on implementation: 

The general thing we’re trying to constrain here is that we evaluate fusions based on profitability. Prologues are just fusions that are less profitable than epilogues. If instead of separately maintaining `prologue/epilogue` dicts, we instead had `dependent_pending_fusions[node1][(node1, node2)]`. Then we could unify the prologue/epilogue loop that you have. 
 In each loop: 
- Resolve the current, pending fusions with `as_completed`. After you resolve a fusion (node1, node2), kick off the next pending fusion from either node1 or node2 based on original ordering of the dependent fusions.
- Continue until done. 

```
This allows you to have a single loop for prologue/epilogue, and it also resolves your TODO: ```
                # TODO (PaulZhang12): Does not support fusions of templates with
                # multiple potential epilogues
```

Anyway, we’ve gone back and forth on a few iterations, and this is an improvement, so happy to land. But it would be great to benchmark the simpler version and ask claude to try out the other proposal to see if it cleans up the TODOs/multiple loops.

PaulZhang12 commented:
@eellison thanks for the detailed response!

Regarding the strawman proposal:
Yes I think this works if we assume generally that things are simple, 1 epilogue per template, a template only has 1 potential fusion. However, in the cases of multiple potential prologues/epilogues for a template, this breaks down/does not fully allow overlap, as you will be evaluating `pending_fusions` immediately in the prior loop.

I can try to simplify the design a bit more. Regarding state tracking/unifying prologue/epilogues, it can get quite tricky and brittle I've found (I had this in previous versions/unified with pending_fusions). I feel like separating them out makes the code a lot more readable generally

PaulZhang12 committed:
[Inductor] Finely overlap template fusion with async_compile (#170565)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/170565

Test Plan: buck run mode/opt caffe2/test/inductor:max_autotune -- -r test_lazy_template_fusion_multiple_candidates 2>&1 | tee ~/test.log

Differential Revision: D89255518
